text/0000-transaction-refactoring.md

```diff
@@ -0,0 +1,327 @@
+We propose to support this type of transaction batching to simplify the runtime.
+
+This change
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+### New transaction and receipts
+
+Previously, in the runtime to produce a block we first executed new signed transactions and then executed received receipts. It resulted in duplicated code that might be shared across similar actions, e.g. function calls for async calls, callbacks and self-calls.
+It also increased the complexity of the runtime implementation.
+
+This NEP proposes changing it by first converting all signed transactions into receipts and then either execute them immediately before received receipts, or put them into the list of the new receipts to be routed.
+To achieve this, NEP introduces a new message `Action` that represents one of atomic actions, e.g. a function call.
+`TransactionBody` is now called just `Transaction`. It contains the list of actions that needs to be performed in a single batch and the information shared across these actions.
+
+`Transaction` contains the following fields
+- `originator_id` is an account ID of the transaction originator.
+- `public_key` is to identify the access key used to signs the transaction.
+- `nonce` is used to deduplicate and order transactions.
+- `receiver_id` is where the transaction has to be routed.
+- `action` is the list of actions to perform.
+
+An `Action` can be of the following:
+- `CreateAccount` creates a new account. It has to be the first action, and the receiver has to be a new account. The action will fail if the account already exists.
+`CreateAccount` also grants permission for all subsequent batched action for the newly created account. For example, deploying code on the new account.
+- `DeployContract` deploys given binary wasm code on the account. Either the `receiver_id` equals to the `originator_id`, or the batch of actions has started with `CreateAccount`, which granted that permission.
+- `FunctionCall` executes a function call on the last deployed contract. E.g. if the previous action was `DeployContract`, then the code to execute will be the new deployed contract. `FunctionCall` has `method_name` and `args` to identify method with arguments to call. It also has `fee` and the `amount`. `fee` is the transaction fee that is prepaid for this call to be spent on gas. `amount` is the attached amount of NEAR tokens that the contract can spend, e.g. 10 tokens to pay for a crypto-corgi.
+- `Transfer` transfers the given amount of tokens from the originator to the receiver.
+- `Stake` stakes the new total amount which is given in the `amount` field with the given `public_key`. The difference in stake is taken from the account's balance at the moment when this action is executed, so it's not prepaid. There is no particular reason to stake on behalf of a newly created account, so we may disallow it.
```

Comment:
The difference is only taken from the account's balance if the transaction tries to increase the stake.

Reply:
Either way it's not prepaid.